src/amr/data/field/coarsening/magnetic_field_coarsener.hpp

```diff
@@ -31,110 +31,216 @@ using core::dirZ;
+template<std::size_t D, typename FieldT>
+typename std::enable_if<D == 1, void>::type
+MagneticFieldCoarsener<dim>::coarsen(Point_t const fineStartIndex, FieldT const& fineField,
+                                     FieldT& coarseField, Point_t const coarseIndex)
+{
+    if (centering_[dirX] == core::QtyCentering::primal) // bx
+    {
+        coarseField(coarseIndex[dirX]) = fineField(fineStartIndex[dirX]);
+    }
+    else if (centering_[dirX] == core::QtyCentering::dual) // by and bz
+    {
+        coarseField(coarseIndex[dirX])
+            = 0.5 * (fineField(fineStartIndex[dirX] + 1) + fineField(fineStartIndex[dirX]));
+    }
+}
+
+template<std::size_t dim>
+template<std::size_t D, typename FieldT>
+typename std::enable_if<D == 2, void>::type
+MagneticFieldCoarsener<dim>::coarsen(Point_t const fineStartIndex, FieldT const& fineField,
+                                     FieldT& coarseField, Point_t const coarseIndex)
+{
+    if (centering_[dirX] == core::QtyCentering::primal
+        and centering_[dirY] == core::QtyCentering::dual)
+    {
+        coarseField(coarseIndex[dirX], coarseIndex[dirY])
+            = 0.5
+              * (fineField(fineStartIndex[dirX], fineStartIndex[dirY])
+                 + fineField(fineStartIndex[dirX], fineStartIndex[dirY] + 1));
+    }
+    else if (centering_[dirX] == core::QtyCentering::dual
+             and centering_[dirY] == core::QtyCentering::primal)
+    {
+        coarseField(coarseIndex[dirX], coarseIndex[dirY])
+            = 0.5
+              * (fineField(fineStartIndex[dirX], fineStartIndex[dirY])
+                 + fineField(fineStartIndex[dirX] + 1, fineStartIndex[dirY]));
+    }
+    else if (centering_[dirX] == core::QtyCentering::dual
+             and centering_[dirY] == core::QtyCentering::dual)
+    {
+        coarseField(coarseIndex[dirX], coarseIndex[dirY])
+            = 0.25
+              * (fineField(fineStartIndex[dirX], fineStartIndex[dirY])
+                 + fineField(fineStartIndex[dirX] + 1, fineStartIndex[dirY])
+                 + fineField(fineStartIndex[dirX], fineStartIndex[dirY] + 1)
+                 + fineField(fineStartIndex[dirX] + 1, fineStartIndex[dirY] + 1));
+    }
+    else
+    {
+        throw std::runtime_error("no magnetic field should end up here");
+    }
+}
+
+// clang-format off
+
+// clang-format on
+
+template<std::size_t dim>
+template<std::size_t D, typename FieldT>
+typename std::enable_if<D == 3, void>::type
+MagneticFieldCoarsener<dim>::coarsen(Point_t const fineStartIndex, FieldT const& fineField,
+                                     FieldT& coarseField, Point_t const coarseIndex)
+{
+    auto& ci  = coarseIndex;
+    auto& cf  = coarseField;
+    auto& fsi = fineStartIndex;
+    auto& ff  = fineField;
+
+    using Fn     = std::function<void()>;
+    using Fn_arr = std::array<Fn, 2>;
+    using A0_arr = std::array<Fn_arr, 2>;
+
+    // X = 0 = primal
+    // Y = 1 = dual
+    std::array<A0_arr, 2> const fns{
```

Comment:
this function is not coarsening correctly the magnetic field.
Magnetic components are on faces. in 3D coarse B compoennts should always take 4 fine cells divided by 1/4. There are only 6 faces and nothing is pure dual.

Reply:
I tried to extrapolate the 2d function without really knowing what I was doing (how could I?)